File: src/image_processor.py
from pathlib import Path
from PIL import Image
from PIL.ExifTags import TAGS, GPSTAGS

from src.face_identifier import FaceIdentifier
from src.scene_classifier import SceneClassifier
from src.object_detector import ObjectDetector
from config import SUPPORTED_FORMATS


class ImageProcessor:
    def __init__(self):
        self.face_identifier = FaceIdentifier()
        self.scene_classifier = SceneClassifier()
        self.object_detector = ObjectDetector()

    def analyze_image(self, image_path):
        image_path = Path(image_path)
        results = {
            "path": str(image_path),
            "filename": image_path.name,
            "faces": self.face_identifier.detect_faces(image_path),
            "scene": self.scene_classifier.classify(image_path),
            "objects": self.object_detector.detect(image_path),
            "metadata": self._extract_metadata(image_path),
        }
        return results

    def _extract_metadata(self, image_path):
        metadata = {
            "date": None,
            "gps_lat": None,
            "gps_lon": None,
            "camera": None,
            "width": None,
            "height": None,
        }

        try:
            img = Image.open(image_path)
            metadata["width"] = img.width
            metadata["height"] = img.height

            if hasattr(img, "_getexif") and img._getexif():
                exif = img._getexif()
                for tag_id, value in exif.items():
                    tag = TAGS.get(tag_id, tag_id)
                    if tag == "DateTimeOriginal":
                        metadata["date"] = str(value)
                    elif tag == "Model":
                        metadata["camera"] = str(value)
                    elif tag == "GPSInfo":
                        gps_data = self._parse_gps(value)
                        metadata["gps_lat"] = gps_data.get("lat")
                        metadata["gps_lon"] = gps_data.get("lon")
        except Exception:
            pass

        return metadata

    def _parse_gps(self, gps_info):
        result = {"lat": None, "lon": None}
        try:
            for key, val in gps_info.items():
                decode = GPSTAGS.get(key, key)
                if decode == "GPSLatitude":
                    result["lat"] = self._convert_to_degrees(val)
                elif decode == "GPSLatitudeRef":
                    if val == "S":
                        result["lat"] = -abs(result["lat"]) if result["lat"] else None
                elif decode == "GPSLongitude":
                    result["lon"] = self._convert_to_degrees(val)
                elif decode == "GPSLongitudeRef":
                    if val == "W":
                        result["lon"] = -abs(result["lon"]) if result["lon"] else None
        except Exception:
            pass
        return result

    def _convert_to_degrees(self, value):
        try:
            d = float(value[0])
            m = float(value[1])
            s = float(value[2])
            return d + (m / 60.0) + (s / 3600.0)
        except Exception:
            return None

    def is_supported(self, file_path):
        return Path(file_path).suffix.lower() in SUPPORTED_FORMATS

File: flutter_photo_organizer/lib/core/services/image_processor.dart
// Composite Image Processor that combines all ML services

import 'dart:io';
import 'package:image/image.dart' as img;
import 'package:path/path.dart' as p;
import 'ml_interfaces.dart';
import '../models/photo_models.dart';
import 'tflite_face_detector.dart';
import 'tflite_face_recognizer.dart';
import 'tflite_scene_classifier.dart';
import 'tflite_object_detector.dart';

class ImageProcessorImpl implements ImageProcessor {
  final FaceDetector _faceDetector;
  final FaceRecognizer _faceRecognizer;
  final SceneClassifier _sceneClassifier;
  final ObjectDetector _objectDetector;
  
  static const List<String> _supportedFormats = [
    '.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp', '.heic', '.tiff'
  ];

  ImageProcessorImpl({
    FaceDetector? faceDetector,
    FaceRecognizer? faceRecognizer,
    SceneClassifier? sceneClassifier,
    ObjectDetector? objectDetector,
  }) : _faceDetector = faceDetector ?? TFLiteFaceDetector(),
       _faceRecognizer = faceRecognizer ?? TFLiteFaceRecognizer(),
       _sceneClassifier = sceneClassifier ?? TFLiteSceneClassifier(),
       _objectDetector = objectDetector ?? TFLiteObjectDetector();

  @override
  Future<ImageAnalysis> analyzeImage(String imagePath) async {
    try {
      final image = await _loadImage(imagePath);
      if (image == null) {
        throw Exception('Failed to load image: $imagePath');
      }

      // Run all analyses in parallel
      final futures = await Future.wait([
        _faceDetector.detectFaces(image),
        _sceneClassifier.classify(image),
        _objectDetector.detect(image),
        _extractMetadata(imagePath, image),
      ]);

      final faces = futures[0] as List<FaceDetection>;
      final scenes = futures[1] as List<SceneClassification>;
      final objects = futures[2] as List<ObjectDetection>;
      final metadata = futures[3] as ImageMetadata;

      // Identify faces
      final identifiedFaces = <FaceDetection>[];
      for (final face in faces) {
        if (face.bbox[2] > face.bbox[0] && face.bbox[3] > face.bbox[1]) {
          final faceCrop = _cropFace(image, face.bbox);
          final name = await _faceRecognizer.identifyFace(faceCrop);
          identifiedFaces.add(FaceDetection(
            bbox: face.bbox,
            confidence: face.confidence,
            name: name,
          ));
        } else {
          identifiedFaces.add(face);
        }
      }

      return ImageAnalysis(
        path: imagePath,
        filename: p.basename(imagePath),
        faces: identifiedFaces,
        scenes: scenes,
        objects: objects,
        metadata: metadata,
      );
    } catch (e) {
      return ImageAnalysis(
        path: imagePath,
        filename: p.basename(imagePath),
        faces: [],
        scenes: [],
        objects: [],
        metadata: ImageMetadata(),
      );
    }
  }

  Future<img.Image?> _loadImage(String path) async {
    try {
      final bytes = await _readFileBytes(path);
      if (bytes == null) return null;
      return img.decodeImage(bytes);
    } catch (e) {
      return null;
    }
  }

  Future<List<int>?> _readFileBytes(String path) async {
    try {
      final file = File(path);
      if (await file.exists()) {
        return await file.readAsBytes();
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  Future<ImageMetadata> _extractMetadata(String imagePath, img.Image image) async {
    // Extract EXIF metadata
    // This would use exif package in a real implementation
    return ImageMetadata(
      width: image.width,
      height: image.height,
      // date, gps, camera would be extracted from EXIF
    );
  }

  img.Image _cropFace(img.Image image, List<int> bbox) {
    final x1 = bbox[0].clamp(0, image.width);
    final y1 = bbox[1].clamp(0, image.height);
    final x2 = bbox[2].clamp(0, image.width);
    final y2 = bbox[3].clamp(0, image.height);
    
    if (x2 <= x1 || y2 <= y1) {
      return img.Image(width: 112, height: 112);
    }
    
    return img.copyCrop(image, x: x1, y: y1, width: x2 - x1, height: y2 - y1);
  }

  @override
  bool isSupported(String filePath) {
    final ext = p.extension(filePath).toLowerCase();
    return _supportedFormats.contains(ext);
  }

  // Getters for sub-services (for advanced usage)
  FaceDetector get faceDetector => _faceDetector;
  FaceRecognizer get faceRecognizer => _faceRecognizer;
  SceneClassifier get sceneClassifier => _sceneClassifier;
  ObjectDetector get objectDetector => _objectDetector;

  void dispose() {
    _faceDetector.dispose();
    _faceRecognizer.dispose();
    _sceneClassifier.dispose();
    _objectDetector.dispose();
  }
}